clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java

```diff
@@ -0,0 +1,245 @@
+        if (coordinators.size() != 1) {
+            log.error(
+                    "Group coordinator lookup failed: Invalid response containing more than a single coordinator");
+            enqueueErrorEvent(new IllegalStateException(
+                    "Group coordinator lookup failed: Invalid response containing more than a single coordinator"));
+        }
+        FindCoordinatorResponseData.Coordinator coordinatorData = coordinators.get(0);
+        Errors error = Errors.forCode(coordinatorData.errorCode());
+        if (error == Errors.NONE) {
+            // use MAX_VALUE - node.id as the coordinator id to allow separate connections
+            // for the coordinator in the underlying network client layer
+            int coordinatorConnectionId = Integer.MAX_VALUE - coordinatorData.nodeId();
+
+            this.coordinator = new Node(
+                    coordinatorConnectionId,
+                    coordinatorData.host(),
+                    coordinatorData.port());
+            log.info("Discovered group coordinator {}", coordinator);
+            coordinatorRequestState.reset();
+            return;
+        }
+
+        if (error == Errors.GROUP_AUTHORIZATION_FAILED) {
+            enqueueErrorEvent(GroupAuthorizationException.forGroupId(this.groupId.orElse(null)));
+            return;
+        }
+
+        log.debug("Group coordinator lookup failed: {}", coordinatorData.errorMessage());
+        enqueueErrorEvent(error.exception());
+    }
+
+    void handleFailedCoordinatorResponse(FindCoordinatorResponse response) {
+        log.debug("FindCoordinator request failed due to {}", response.error().toString());
+
+        if (!(response.error().exception() instanceof RetriableException)) {
+            log.info("FindCoordinator request hit fatal exception", response.error().exception());
+            // Remember the exception if fatal so we can ensure
+            // it gets thrown by the main thread
+            enqueueErrorEvent(response.error().exception());
+        }
+
+        log.debug("Coordinator discovery failed, refreshing metadata", response.error().exception());
+    }
+
+    /**
+     * Handle 3 cases of FindCoordinatorResponse: success, failure, timedout
+     *
+     * @param response FindCoordinator response
+     */
+    public void onResponse(FindCoordinatorResponse response) {
+        coordinatorRequestState.updateLastReceived();
+        if (response.hasError()) {
+            handleFailedCoordinatorResponse(response);
+            return;
+        }
+        handleSuccessFindCoordinatorResponse(response);
+    }
+
+    private void enqueueErrorEvent(Exception e) {
+        backgroundEventQueue.add(new ErrorBackgroundEvent(e));
+    }
+
+    public Node coordinator() {
+        return coordinator;
+    }
+
+    private class CoordinatorRequestState {
+        private long lastSentMs;
+        private long lastReceivedMs;
+        private int numAttempts;
+        public CoordinatorRequestState() {
```

Comment:
nit: we can leave out the constructor and initialize the fields directly?

Reply:
will move it to the findCoordinator method.

Reply:
I just mean we can do this:
```java
        private long lastSentMs = -1;
        private long lastReceivedMs = -1;
        private int numAttempts = 0;
```